### PR TITLE
Support symlinking of the /pub document root

### PR DESCRIPTION
**Pull request by gwillem**
Magento recommends to use `/pub` as document root. Sometimes, the document root is at a fixed path. Given this installation where `/var/www` is the configured document root:

```
/data/magento2-repo-checkout
/var/www (symlinks to) => /data/magento2-repo-checkout/pub
```

It will break:

```
PHP Fatal error:  require(): Failed opening required '/var/www/../app/bootstrap.php' (include_path='.:/usr/share/php:/usr/share/pear') in /var/www/index.php on line 13
```

..because relative paths (`../`) assume `realpath()`

With this PR, top level symlinks to `pub` will work as expected.

### CODE REVIEW

**Comment by allardhoeve** [comment]
Ohh, I want this. :+1:

**Comment by genisd** [comment]
me too :hushed:

**Comment by JeroenBoersma** [comment]
+1

**Comment by yireo** [comment]
+1

**Comment by simkea** [comment]
i need also a general symlink support
+1 ;)

**Comment by JeroenvHeugten** [comment]
:+1:

**Comment by vancoz** [comment]
Hello @gwillem thank you for your contribution, internal ticket number is MAGETWO-48217

**Comment by tkn98** [comment]
Please take care for the inverse scenario as well:

`/data/magento2-repo-checkout/pub` symlinks to `var/www`.

Most often such problems occur if there is too much use of realpath() in the wrong places on the wrong paths. path resolution often isn't real-path resolution.

Just my 2 cents.

**Comment by jasperzeinstra** [comment]
+1

**Comment by genisd** [comment]
> Please take care for the inverse scenario as well:
> /data/magento2-repo-checkout/pub symlinks to var/www.

Just tested this. Line 13 (in this PR) in `pub/index.php` will not work with this right now.

The setup: `/data/web/magento2/pub/` symlinking to `/data/web/public/` (this is also the webserver root).

This will yield in an import attempt like this: `require(): Failed opening required '/data/web/public/../app/bootstrap.php'`

To me it appears difficult to support both scenarios. In my opinion just one of the two scenarios should work, such that everyone can enjoy the flexibility that a symlink can provide in this scenario.

**Comment by tkn98** [comment]
@genisd: Yes for some reason there are made hidden assumptions / dependencies on the realpath. That causes errors on "best-guessed" relative path resolutions. IIRC this has been reported earlier, but I don't have a ticket at hand nor know about the resolution. If the overall then is well documented and some rough edges can be removed along the line I'd say this will benefit everybody (even if not all scenarios are supported in the end).

**Comment by arkadiych** [comment]
@genisd and @gwillem, you should sign Contributor License Agreement

**Comment by genisd** [comment]
I have. :( Let me try again
I just tried again with another browser. Perhaps @gwillem needs to re-sign it again or something?

**Comment by gwillem** [comment]
@arkadiych I just did

**Comment by okorshenko** [comment]
@gwillem your Pull Request merged to develop branch. Thank you for contribution!

**Comment by mazhalai** [comment]
@gwillem Thank you for your contribution, your PR has been merged.
